vyper/venom/analysis/available_expression.py

```diff
@@ -0,0 +1,272 @@
+            if isinstance(op, _Expression):
+                d = op.get_depth()
+                if d > max_depth:
+                    max_depth = d
+        return max_depth + 1
+
+    @cached_property
+    def get_reads(self) -> Effects:
+        tmp_reads = self.first_inst.get_read_effects()
+        for op in self.operands:
+            if isinstance(op, _Expression):
+                tmp_reads = tmp_reads | op.get_reads
+        if self.ignore_msize:
+            tmp_reads &= ~Effects.MSIZE
+        return tmp_reads
+
+    @cached_property
+    def get_writes(self) -> Effects:
+        tmp_reads = self.first_inst.get_write_effects()
+        for op in self.operands:
+            if isinstance(op, _Expression):
+                tmp_reads = tmp_reads | op.get_writes
+        if self.ignore_msize:
+            tmp_reads &= ~Effects.MSIZE
+        return tmp_reads
+
+    @property
+    def is_commutative(self) -> bool:
+        return self.first_inst.is_commutative
+
+
+class _BBLattice:
+    data: dict[IRInstruction, OrderedSet[_Expression]]
+    out: OrderedSet[_Expression]
+
+    def __init__(self, bb: IRBasicBlock):
+        self.data = dict()
+        self.out = OrderedSet()
+        for inst in bb.instructions:
+            self.data[inst] = OrderedSet()
+
+
+class _FunctionLattice:
+    data: dict[IRBasicBlock, _BBLattice]
+
+    def __init__(self, function: IRFunction):
+        self.data = dict()
+        for bb in function.get_basic_blocks():
+            self.data[bb] = _BBLattice(bb)
+
+
+class AvailableExpressionAnalysis(IRAnalysis):
+    expressions: OrderedSet[_Expression] = OrderedSet()
+    inst_to_expr: dict[IRInstruction, _Expression] = dict()
+    dfg: DFGAnalysis
+    lattice: _FunctionLattice
+
+    # the size of the expressions
+    # that are considered in the analysis
+    min_depth: int
+    max_depth: int
+    ignore_msize: bool
+
+    def __init__(
+        self,
+        analyses_cache: IRAnalysesCache,
+        function: IRFunction,
+        min_depth: int = _MIN_DEPTH,
+        max_depth: int = _MAX_DEPTH,
+    ):
+        super().__init__(analyses_cache, function)
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        dfg = self.analyses_cache.request_analysis(DFGAnalysis)
+        assert isinstance(dfg, DFGAnalysis)
+        self.dfg = dfg
+
+        self.min_depth = min_depth
+        self.max_depth = max_depth
+
+        self.lattice = _FunctionLattice(function)
+
+        self.ignore_msize = not self._contains_msize()
+
+    def analyze(self, min_depth: int = _MIN_DEPTH, max_depth: int = _MAX_DEPTH):
+        self.min_depth = min_depth
+        self.max_depth = max_depth
+
+        worklist: deque = deque()
+        worklist.append(self.function.entry)
+        while len(worklist) > 0:
+            bb: IRBasicBlock = worklist.popleft()
+            changed = self._handle_bb(bb)
+
+            if changed:
+                for out in bb.cfg_out:
+                    if out not in worklist:
```

Comment:
if we require uniqueness in the worklist then we should use OrderedSet. it has a `pop()` function which is O(1) LIFO pop.